### PR TITLE
Support alphanumeric sort in search query

### PR DESCRIPTION
**Pull request by jaehc**
I did the following things.

- support alphanumeric sort in a search query.
- place the same string comparison functions of groupby and search in one class. but there is a slight difference from previous one in functionality like null first.
- support strlen sort for groupby as a side effect.
- update docs about sorting in grouby/search query.

### CODE REVIEW

**Comment by navis** [comment]
equals/hashCode are missing in search sort specs. It was just like that before this PR but it would be good to fix it here altogether.

**Comment by CHOIJAEHONG1** [comment]
rebased and merge conflicts

File: processing/src/main/java/io/druid/query/search/search/LexicographicSearchSortSpec.java [hunk 1]
```diff
@@ -63,6 +68,12 @@ public String toString()
 
   @Override
   public boolean equals(Object other) {
-    return (other instanceof LexicographicSearchSortSpec);
+    return this == other || other instanceof LexicographicSearchSortSpec;
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/search/search/LexicographicSearchSortSpec.java]: can we please auto generate these?

File: processing/src/main/java/io/druid/query/search/search/AlphanumericSearchSortSpec.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  public String toString()
+  {
+    return "alphanumericSort";
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    return this == other || other instanceof AlphanumericSearchSortSpec;
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/search/search/AlphanumericSearchSortSpec.java]: please auto-generate these

File: processing/src/main/java/io/druid/query/ordering/StringComparators.java [hunk 3]
```diff
@@ -293,12 +297,57 @@ public String toString()
...
+    }
+
+    @Override
+    public String toString()
+    {
+      return StringComparators.STRLEN_NAME;
+    }
+  }
+
   public static StringComparator makeComparator(String type)
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/ordering/StringComparators.java]: why is this method needed? can we use polymorphic deserialization instead?

File: processing/src/main/java/io/druid/query/ordering/StringComparators.java [hunk 4]
```diff
@@ -293,12 +297,57 @@ public String toString()
     }
   }
 
+  public static class StrlenComparator implements StringComparator
+  {
+    @Override
+    public int compare(String s, String s2)
+    {
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/ordering/StringComparators.java]: can we just use Ordering.natural().nullFirst9).compare?

File: processing/src/main/java/io/druid/query/ordering/StringComparators.java (hunk 3 above)
Comment (CHOIJAEHONG1) [inline processing/src/main/java/io/druid/query/ordering/StringComparators.java]: This method is used by OrderByColumnSpec for groupby queries, making instances from strings of a Map object, as in.
https://github.com/druid-io/druid/blob/master/processing/src/main/java/io/druid/query/groupby/orderby/OrderByColumnSpec.java#L72

IMHO, this appears to be changed beforehand.

File: processing/src/main/java/io/druid/query/search/search/LexicographicSearchSortSpec.java (hunk 1 above)
Comment (CHOIJAEHONG1) [inline processing/src/main/java/io/druid/query/search/search/LexicographicSearchSortSpec.java]: Do you mean IDE-generated one? My IDE(eclipse Mars) could not generate `equals()` automatically becasue these classes do not have non-static fields. I copied `equals()` from `LexicographicSearchSortSpec` which dates back to [](https://github.com/druid-io/druid/commit/4c8495662f672f146bd1f7eb64cd3bd3096a7aed#diff-e3e13007b63516c87e96b8cfc1ae8a40)

File: processing/src/main/java/io/druid/query/ordering/StringComparators.java [hunk 5]
```diff
@@ -282,23 +291,66 @@ public boolean equals(Object o)
...
+  {
+    private static final Ordering<String> ORDERING = Ordering.from(new Comparator<String>()
+      {
+        @Override
+        public int compare(String s, String s2)
+        {
+          return Ints.compare(s.length(), s2.length());
+        }
+      }).nullsFirst().compound(Ordering.natural());
+
```
Comment (navis) [inline processing/src/main/java/io/druid/query/ordering/StringComparators.java]: spaces

File: processing/src/main/java/io/druid/query/ordering/StringComparators.java [hunk 6]
```diff
@@ -34,40 +36,44 @@
...
+      {
+        @Override
+        public int compare(String s, String s2)
+        {
+          return UnsignedBytes.lexicographicalComparator().compare(
+              StringUtils.toUtf8(s),
+              StringUtils.toUtf8(s2));
+        }
+      }).nullsFirst();
+
```
Comment (navis) [inline processing/src/main/java/io/druid/query/ordering/StringComparators.java]: spaces

File: processing/src/main/java/io/druid/query/ordering/StringComparators.java (hunk 5 above)
Comment (CHOIJAEHONG1) [inline processing/src/main/java/io/druid/query/ordering/StringComparators.java]: done.

File: processing/src/main/java/io/druid/query/ordering/StringComparators.java (hunk 6 above)
Comment (CHOIJAEHONG1) [inline processing/src/main/java/io/druid/query/ordering/StringComparators.java]: done.

**Comment by navis** [comment]
:+1: when test passes

**Comment by fjy** [comment]
👍

**Pull request merged**